File: apps/macos/Kelpie/Handlers/Snapshot3DHandler.swift
import Foundation

struct Snapshot3DHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("snapshot-3d-enter") { _ in await enter() }
        router.register("snapshot-3d-exit") { _ in await exit() }
        router.register("snapshot-3d-status") { _ in await status() }
        router.register("snapshot-3d-set-mode") { body in await setMode(body) }
        router.register("snapshot-3d-zoom") { body in await zoom(body) }
        router.register("snapshot-3d-reset-view") { _ in await resetView() }
    }

    @MainActor
    private func enter() async -> [String: Any] {
        guard FeatureFlags.is3DInspectorEnabled else {
            return errorResponse(
                code: "FEATURE_DISABLED",
                message: "3D inspector is not enabled. Enable in Settings or set KELPIE_3D_INSPECTOR=1"
            )
        }
        guard !context.isIn3DInspector else {
            return errorResponse(code: "ALREADY_ACTIVE", message: "3D inspector is already active")
        }

        do {
            try await context.evaluateJS(Snapshot3DBridge.enterScript)
            let active = try await context.evaluateJSReturningString("window.__m3d ? 'true' : 'false'")
            if active == "true" {
                context.isIn3DInspector = true
                return successResponse()
            }
            return errorResponse(code: "ACTIVATION_FAILED", message: "3D inspector script did not activate")
        } catch {
            return errorResponse(code: "JS_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func exit() async -> [String: Any] {
        guard context.isIn3DInspector else {
            return successResponse()
        }

        do {
            try await context.evaluateJS(Snapshot3DBridge.exitScript)
            context.mark3DInspectorInactive(notify: true)
            return successResponse()
        } catch {
            return errorResponse(code: "JS_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func status() async -> [String: Any] {
        successResponse(["active": context.isIn3DInspector])
    }

    @MainActor
    private func setMode(_ body: [String: Any]) async -> [String: Any] {
        guard context.isIn3DInspector else {
            return errorResponse(code: "NOT_ACTIVE", message: "3D inspector is not active")
        }
        let requested = (body["mode"] as? String ?? "rotate").lowercased()
        guard requested == "rotate" || requested == "scroll" else {
            return errorResponse(code: "INVALID_MODE", message: "mode must be 'rotate' or 'scroll'")
        }
        do {
            let applied = try await context.evaluateJSReturningString(Snapshot3DBridge.setModeScript(requested))
            let normalized = applied == "rotate" || applied == "scroll" ? applied : requested
            return successResponse(["mode": normalized])
        } catch {
            return errorResponse(code: "JS_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func zoom(_ body: [String: Any]) async -> [String: Any] {
        guard context.isIn3DInspector else {
            return errorResponse(code: "NOT_ACTIVE", message: "3D inspector is not active")
        }
        let delta: Double
        if let d = body["delta"] as? Double {
            delta = d
        } else if let direction = body["direction"] as? String {
            switch direction.lowercased() {
            case "in": delta = 0.1
            case "out": delta = -0.1
            default:
                return errorResponse(code: "INVALID_DIRECTION", message: "direction must be 'in' or 'out'")
            }
        } else {
            return errorResponse(code: "MISSING_PARAM", message: "Provide 'delta' (number) or 'direction' ('in'|'out')")
        }
        do {
            try await context.evaluateJS(Snapshot3DBridge.zoomByScript(delta))
            return successResponse(["delta": delta])
        } catch {
            return errorResponse(code: "JS_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func resetView() async -> [String: Any] {
        guard context.isIn3DInspector else {
            return errorResponse(code: "NOT_ACTIVE", message: "3D inspector is not active")
        }
        do {
            try await context.evaluateJS(Snapshot3DBridge.resetViewScript)
            return successResponse()
        } catch {
            return errorResponse(code: "JS_ERROR", message: error.localizedDescription)
        }
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/Snapshot3DHandler.kt
package com.kelpie.browser.handlers

class Snapshot3DHandler(
    private val context: HandlerContext,
    private val featureFlagEnabled: () -> Boolean,
) {
    fun register(router: com.kelpie.browser.network.Router) {
        router.register("snapshot-3d-enter") { enter() }
        router.register("snapshot-3d-exit") { exit() }
        router.register("snapshot-3d-status") { status() }
        router.register("snapshot-3d-set-mode") { body -> setMode(body) }
        router.register("snapshot-3d-zoom") { body -> zoom(body) }
        router.register("snapshot-3d-reset-view") { resetView() }
    }

    private suspend fun enter(): Map<String, Any?> {
        if (!featureFlagEnabled()) {
            return com.kelpie.browser.network.errorResponse(
                "FEATURE_DISABLED",
                "3D inspector is disabled in Settings",
            )
        }
        if (context.isIn3DInspector) {
            return com.kelpie.browser.network
                .errorResponse("ALREADY_ACTIVE", "3D inspector is already active")
        }

        return try {
            context.evaluateJS(Snapshot3DBridge.ENTER_SCRIPT)
            val active = context.evaluateJS("!!window.__m3d")
            if (active.contains("true")) {
                context.isIn3DInspector = true
                com.kelpie.browser.network
                    .successResponse()
            } else {
                com.kelpie.browser.network
                    .errorResponse("ACTIVATION_FAILED", "3D inspector script did not activate")
            }
        } catch (error: Exception) {
            com.kelpie.browser.network
                .errorResponse("JS_ERROR", error.localizedMessage ?: "JavaScript error")
        }
    }

    private suspend fun exit(): Map<String, Any?> {
        if (!context.isIn3DInspector) {
            return com.kelpie.browser.network
                .successResponse()
        }

        return try {
            context.evaluateJS(Snapshot3DBridge.EXIT_SCRIPT)
            context.mark3DInspectorInactive()
            com.kelpie.browser.network
                .successResponse()
        } catch (error: Exception) {
            com.kelpie.browser.network
                .errorResponse("JS_ERROR", error.localizedMessage ?: "JavaScript error")
        }
    }

    private fun status(): Map<String, Any?> =
        com.kelpie.browser.network
            .successResponse(mapOf("active" to context.isIn3DInspector))

    private suspend fun setMode(body: Map<String, Any?>): Map<String, Any?> {
        if (!context.isIn3DInspector) {
            return com.kelpie.browser.network
                .errorResponse("NOT_ACTIVE", "3D inspector is not active")
        }
        val requested = (body["mode"] as? String ?: "rotate").lowercase()
        if (requested != "rotate" && requested != "scroll") {
            return com.kelpie.browser.network
                .errorResponse("INVALID_MODE", "mode must be 'rotate' or 'scroll'")
        }
        return try {
            val applied = context.evaluateJS(Snapshot3DBridge.setModeScript(requested))
            val normalized =
                when {
                    applied.contains("rotate") -> "rotate"
                    applied.contains("scroll") -> "scroll"
                    else -> requested
                }
            com.kelpie.browser.network
                .successResponse(mapOf("mode" to normalized))
        } catch (error: Exception) {
            com.kelpie.browser.network
                .errorResponse("JS_ERROR", error.localizedMessage ?: "JavaScript error")
        }
    }

    private suspend fun zoom(body: Map<String, Any?>): Map<String, Any?> {
        if (!context.isIn3DInspector) {
            return com.kelpie.browser.network
                .errorResponse("NOT_ACTIVE", "3D inspector is not active")
        }
        val delta: Double =
            when {
                body["delta"] is Number -> (body["delta"] as Number).toDouble()
                body["direction"] is String ->
                    when ((body["direction"] as String).lowercase()) {
                        "in" -> 0.1
                        "out" -> -0.1
                        else -> return com.kelpie.browser.network.errorResponse(
                            "INVALID_DIRECTION",
                            "direction must be 'in' or 'out'",
                        )
                    }
                else -> return com.kelpie.browser.network.errorResponse(
                    "MISSING_PARAM",
                    "Provide 'delta' (number) or 'direction' ('in'|'out')",
                )
            }
        return try {
            context.evaluateJS(Snapshot3DBridge.zoomByScript(delta))
            com.kelpie.browser.network
                .successResponse(mapOf("delta" to delta))
        } catch (error: Exception) {
            com.kelpie.browser.network
                .errorResponse("JS_ERROR", error.localizedMessage ?: "JavaScript error")
        }
    }

    private suspend fun resetView(): Map<String, Any?> {
        if (!context.isIn3DInspector) {
            return com.kelpie.browser.network
                .errorResponse("NOT_ACTIVE", "3D inspector is not active")
        }
        return try {
            context.evaluateJS(Snapshot3DBridge.RESET_VIEW_SCRIPT)
            com.kelpie.browser.network
                .successResponse()
        } catch (error: Exception) {
            com.kelpie.browser.network
                .errorResponse("JS_ERROR", error.localizedMessage ?: "JavaScript error")
        }
    }
}
